opencensus/exporters/trace/zipkin/internal/zipkin_exporter.cc

```diff
@@ -0,0 +1,361 @@
+
+namespace opencensus {
+namespace exporters {
+namespace trace {
+
+namespace {
+
+constexpr char kZipkinLib[] = "zipkin/2.0";
+constexpr char ipv4_loopback[] = "127.0.0.1";
+constexpr char ipv6_loopback[] = "::1";
+
+std::string AttributeValueToString(
+    const ::opencensus::trace::exporter::AttributeValue &value) {
+  switch (value.type()) {
+    case ::opencensus::trace::AttributeValueRef::Type::kString:
+      return value.string_value();
+      break;
+    case ::opencensus::trace::AttributeValueRef::Type::kBool:
+      return value.bool_value() ? "true" : "false";
+      break;
+    case ::opencensus::trace::AttributeValueRef::Type::kInt:
+      return std::to_string(value.int_value());
+      break;
+  }
+  return "";
+}
+
+std::string SerializeAnnotation(
+    const ::opencensus::trace::exporter::Annotation &annotation) {
+  std::string annotation_str(annotation.description());
+  absl::StrAppend(&annotation_str, " (");
+  size_t count = 0;
+  for (const auto &attribute : annotation.attributes()) {
+    absl::StrAppend(&annotation_str, attribute.first, ":",
+                    AttributeValueToString(attribute.second));
+
+    if (++count < annotation.attributes().size()) {
+      absl::StrAppend(&annotation_str, ", ");
+    }
+  }
+  absl::StrAppend(&annotation_str, ")");
+  return annotation_str;
+}
+
+void SerializeJson(const ::opencensus::trace::exporter::SpanData &span,
+                   const ZipkinExporterOptions::Service &service,
+                   rapidjson::Writer<rapidjson::StringBuffer> *writer) {
+  writer->StartObject();
+
+  writer->Key("name");
+  writer->String(std::string(span.name()));
+
+  writer->Key("traceId");
+  writer->String(span.context().trace_id().ToHex());
+
+  writer->Key("id");
+  writer->String(span.context().span_id().ToHex());
+
+  if (span.parent_span_id().IsValid()) {
+    writer->Key("parentId");
+    writer->String(span.parent_span_id().ToHex());
+  }
+
+  // Write endpoint.  Census does not support this by default.
+  writer->Key("localEndpoint");
+  writer->StartObject();
+  writer->Key("serviceName");
+  writer->String(service.service_name);
+  if (service.af_type == ZipkinExporterOptions::AddressFamily::kIpv6) {
+    writer->Key("ipv6");
+    writer->String(service.ip_address);
+  } else {
+    writer->Key("ipv4");
+    writer->String(service.ip_address);
+  }
+  writer->EndObject();
+
+  if (!span.annotations().events().empty()) {
+    writer->Key("annotations");
+    writer->StartArray();
+    for (const auto &annotation : span.annotations().events()) {
+      writer->StartObject();
+      writer->Key("timestamp");
+      writer->Int64(absl::ToUnixMicros(annotation.timestamp()));
+      writer->Key("value");
+      writer->String(SerializeAnnotation(annotation.event()));
+      writer->EndObject();
+    }
+    writer->EndArray(span.annotations().events().size());
+  }
+
+  if (!span.attributes().empty()) {
+    writer->Key("tags");
```

Comment:
tags are not an array, they are a dict.

{ "foo": "bar" }

not
[ { "key": "foo", "value": "bar" } ]

Reply:
Fixed.